irohad/consensus/yac/impl/yac_gate_impl.cpp

```diff
@@ -48,104 +39,102 @@ namespace iroha {
-                rxcpp::observable<>::iterate(commit_message.votes)
-                    // allow other peers to apply commit
-                    .flat_map([this, model_hash](auto vote) {
-                      // map vote to block if it can be loaded
-                      return rxcpp::observable<>::create<
-                          std::shared_ptr<shared_model::interface::Block>>(
-                          [this, model_hash, vote](auto subscriber) {
-                            auto block = block_loader_->retrieveBlock(
-                                vote.signature->publicKey(),
-                                shared_model::crypto::Hash(model_hash));
-                            // if load is successful
-                            if (block) {
-                              // update the cache with block consensus voted for
-                              consensus_result_cache_->insert(*block);
-                              subscriber.on_next(*block);
-                            } else {
-                              log_->error(
-                                  "Could not get block from block loader");
-                            }
-                            subscriber.on_completed();
-                          });
-                    })
-                    // need only the first
-                    .first()
-                    .retry()
-                    .subscribe(
-                        // if load is successful from at least one node
-                        [subscriber](auto block) {
-                          subscriber.on_next(block);
-                          subscriber.on_completed();
-                        },
-                        // if load has failed, no peers provided the block
-                        [this, subscriber](std::exception_ptr) {
-                          log_->error("Cannot load committed block");
-                          subscriber.on_completed();
-                        });
-              });
+          return visit_in_place(
+              message,
+              [this](const CommitMessage &msg) { return handleCommit(msg); },
+              [this](const RejectMessage &msg) { return handleReject(msg); });
         });
       }
 
       void YacGateImpl::copySignatures(const CommitMessage &commit) {
         for (const auto &vote : commit.votes) {
           auto sig = vote.hash.block_signature;
-          current_block_.second->addSignature(sig->signedData(),
-                                              sig->publicKey());
+          current_block_->addSignature(sig->signedData(), sig->publicKey());
         }
       }
+
+      rxcpp::observable<YacGateImpl::GateObject> YacGateImpl::handleCommit(
+          const CommitMessage &msg) {
+        return rxcpp::observable<>::create<GateObject>([&](auto subscriber) {
+          const auto hash = getHash(msg.votes);
+          if (not hash) {
+            log_->info("Invalid commit message, hashes are different");
```

Comment:
Commit message cannot be invalid, since it is guaranteed by CommitMessage interface. Consider removing this conditional.